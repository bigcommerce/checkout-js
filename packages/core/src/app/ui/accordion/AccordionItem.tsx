import classNames from 'classnames';
import React, { FunctionComponent, memo, ReactNode, useCallback, useContext } from 'react';
import { CSSTransition } from 'react-transition-group';

import AccordionContext from './AccordionContext';

export interface AccordionItemProps {
    bodyClassName?: string;
    children?: ReactNode;
    className?: string;
    classNameSelected?: string;
    headerClassName?: string;
    headerClassNameSelected?: string;
    itemId: string;
    headerContent(props: AccordionItemHeaderProps): ReactNode;
}

export interface AccordionItemHeaderProps {
    isSelected: boolean;
    onToggle(id: string): void;
}

const AccordionItem: FunctionComponent<AccordionItemProps> = ({
    bodyClassName = 'accordion-item-body',
    children,
    className = 'accordion-item',
    classNameSelected = 'accordion-item--selected',
    headerClassName = 'accordion-item-header',
    headerClassNameSelected = 'accordion-item-header--selected',
    headerContent,
    itemId,
}) => {
    const { onToggle, selectedItemId } = useContext(AccordionContext);
    const isSelected = selectedItemId === itemId;

    const transitionEndListener = useCallback((node, done) => {
        node.addEventListener('transitionend', ({ target }: Event) => {
            if (target === node) {
                done();
            }
        });
    }, []);

    return (
        <li
            className={classNames(className, { [classNameSelected]: isSelected })}
        >
            <div className={classNames(headerClassName, { [headerClassNameSelected]: isSelected })}>
                {headerContent({ isSelected, onToggle })}
            </div>

            {children && (
                <CSSTransition
                    addEndListener={transitionEndListener}
                    classNames={bodyClassName}
                    in={isSelected}
                    mountOnEnter
                    timeout={{}}
                    unmountOnExit
                >
                    <div className={bodyClassName}>{children}</div>
                </CSSTransition>
            )}
        </li>
    );
};

export default memo(AccordionItem);
