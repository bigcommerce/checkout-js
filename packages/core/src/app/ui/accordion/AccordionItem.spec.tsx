import { mount } from 'enzyme';
import React from 'react';

import AccordionItem from './AccordionItem';

describe('AccordionItem', () => {
    it('renders component that matches snapshot', () => {
        const component = mount(
            <AccordionItem headerContent={() => 'Foobar'} itemId="foobar">
                Hello world
            </AccordionItem>,
        );

        expect(component).toMatchSnapshot();
    });

    it('overrides default class names', () => {
        const component = mount(
            <AccordionItem
                className="item"
                headerClassName="header"
                headerContent={() => 'Foobar'}
                itemId="foobar"
            >
                Hello world
            </AccordionItem>,
        );

        expect(component.exists('.item')).toBe(true);

        expect(component.exists('.header')).toBe(true);
    });
});
