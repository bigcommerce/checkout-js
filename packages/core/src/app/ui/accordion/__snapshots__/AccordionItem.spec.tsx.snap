// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem renders component that matches snapshot 1`] = `
<Memo(AccordionItem)
  headerContent={[Function]}
  itemId="foobar"
>
  <li
    className="accordion-item"
  >
    <div
      className="accordion-item-header"
    >
      Foobar
    </div>
    <CSSTransition
      addEndListener={[Function]}
      classNames="accordion-item-body"
      in={false}
      mountOnEnter={true}
      timeout={Object {}}
      unmountOnExit={true}
    >
      <Transition
        addEndListener={[Function]}
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={Object {}}
        unmountOnExit={true}
      />
    </CSSTransition>
  </li>
</Memo(AccordionItem)>
`;
